File: ios/LibsnarkModule.swift
//
//  libsnark_wrapper.swift
//  azerothFront
//
//  Created by Jaekyoung Choi on 2022/06/09.
//

import Foundation
import CSnark

@objc(LibsnarkModule)
class LibsnarkModule: NSObject {
  
  static let serializeFormatDefault : Int32 = 1
  static let serializeFormatCRV : Int32 = 2
  static let serializeFormatZKlay : Int32 = 3
  
  let R1CS_GG : Int32     = 1
  let R1CS_ROM_SE : Int32 = 2
  
  public static let EC_ALT_BN128 : Int32 = 1
  public static let EC_BLS12_381 : Int32 = 2
  
  private var context_id : Int32 = 0
  private var context_id_map: [String: Int32] = ["ZKlay": 0, "ZKlay_nft": 0, "GenTrade":0];
  var contextID : Int { get { return Int(context_id) } }
  
  @objc(getContextId:resolver:rejecter:)
  func getContextId(_ circuitName: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let resolveData: [String: String] = ["contextId": "\(context_id_map[circuitName] ?? 0)"];
    resolve(resolveData)
  }
  
  @objc(setContextId:resolver:rejecter:)
  func setContextId(_ circuitName: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    context_id = context_id_map[circuitName] ?? 0;
    
    let resolveData: [String: String] = ["contextId": "\(context_id_map[circuitName] ?? 0)"];
    resolve(resolveData)
  }
  
  @objc(loggerTest:resolver:rejecter:)
  func loggerTest(_ name: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let str = name
    if (str != "") {
      resolve(str)
    } else {
      let error = NSError(domain: "", code: 200, userInfo: nil)
      reject("event_failure", "no str returned", error)
    }
  }
  
  @objc
  func testMethod() -> Void {
    print("Print test message [libsnark_wrapper.swift] ")
  }
  
  @objc
  func constantsToExport() -> [AnyHashable : Any]! {
    return ["projectName": "ModuleName"]
  }
  
  @objc(createCircuitContext:treeHeight:hashType:serializeFormat:ecSelection:resolver:rejecter:)
  func createCircuitContext(_ circuitName: String, treeHeight: String, hashType: String,  serializeFormat: Int32 = serializeFormatZKlay, ecSelection: Int32 = EC_ALT_BN128, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let _circuitName = CString(circuitName)
    let contextId: Int32 = CSnark.createCircuitContext(_circuitName.char(),
                                                       R1CS_GG,
                                                       ecSelection,
                                                       CString("").char(),
                                                       CString("").char(),
                                                       CString("").char())
    context_id_map[circuitName] = contextId;
    context_id = contextId;
    
    let _treeHeight = CString(treeHeight)
    let _hashType = CString(hashType)
    
    CSnark.serializeFormat(contextId , serializeFormat )
    CSnark.assignCircuitArgument(contextId, CString("treeHeight").char(), _treeHeight.char())
    CSnark.assignCircuitArgument(contextId, CString("hashType").char(), _hashType.char())
    
    let resolveData: [String: String] = ["circuitName": circuitName, "contextId": "\(contextId)"]
    resolve(resolveData)
  }
  
  @objc(createGenTradeCircuitContext:serializeFormat:ecSelection:resolver:rejecter:)
  func createGenTradeCircuitContext(_ circuitName: String, serializeFormat: Int32 = serializeFormatZKlay, ecSelection: Int32 = EC_ALT_BN128, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let _circuitName = CString(circuitName)
    let contextId: Int32 = CSnark.createCircuitContext(_circuitName.char(),
                                                       R1CS_GG,
                                                       ecSelection,
                                                       CString("").char(),
                                                       CString("").char(),
                                                       CString("").char())
    context_id_map[circuitName] = contextId;
    context_id = contextId;
    
    CSnark.serializeFormat(contextId , serializeFormat )
    
    let resolveData: [String: String] = ["circuitName": circuitName, "contextId": "\(contextId)"]
    resolve(resolveData)
  }
  
  @objc(buildCircuit:rejecter:)
  func buildCircuit(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.buildCircuit(context_id)
    if (rtn != 0) {
      reject("event_failure","buildCircuit: \(rtn), \(lastmsg(contextId:context_id))", Errors.buildCircuitError)
    }
    let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
    resolve(resolveData)
  }
  
  @objc(runSetup:rejecter:)
  func runSetup(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.runSetup(context_id);
    if (rtn != 0) {
      reject("event_failure","runSetupError: \(rtn), \(lastmsg(contextId:context_id))", Errors.runSetupError)
    }
    let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
    resolve(resolveData)
  }
  
  @objc(runProof:rejecter:)
  func runProof(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.runProof(context_id)
    if (rtn != 0) {
      reject("event_failure","runSetupError: \(rtn), \(lastmsg(contextId:context_id))", Errors.runProofError)
    }
    let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
    resolve(resolveData)
  }
  
  @objc(runVerify:rejecter:)
  func runVerify(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.runVerify(context_id)
    if (rtn != 0) {
      reject("event_failure","runVerifyError: \(rtn), \(lastmsg(contextId:context_id))", Errors.runProofError)
    }
    let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
    resolve(resolveData)
  }
  
  @objc(setProof:resolver:rejecter:)
  func setProof(proofJsonStr: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.deSerializeProof(context_id, CString(proofJsonStr).char())
    if (rtn != 0) {
      reject("event_failure","runSetupError: \(rtn), \(lastmsg(contextId:context_id))", Errors.runProofError)
    }
    let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
    resolve(resolveData)
  }
  
  @objc(getVerifyKeyJson:rejecter:)
  func getVerifyKeyJson(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    if let Cstr : UnsafePointer<CChar> = CSnark.serializeVerifyKey( context_id ){
      resolve (String(cString: ( Cstr )))
    } else{
      reject("event_failure","serializeVerifyKeyError", Errors.serializeVerifyKeyError)
    }
  }
  
  @objc(getProofJson:rejecter:)
  func getProofJson(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    if let Cstr : UnsafePointer<CChar> = CSnark.serializeProof( context_id ){
      resolve (String(cString: ( Cstr )))
    } else{
      reject("event_failure","serializeProofError", Errors.serializeProofError)
    }
  }
  
  @objc(writeVerifyKeyToFile:resolver:rejecter:)
  func writeVerifyKeyToFile(filePath: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.writeVK( context_id , CString( Bundle.main.bundlePath + filePath).char() )
    if (rtn != 0){
      reject("event_failure","writeVerifyKeyToFile: \(rtn), \(lastmsg(contextId:context_id))", Errors.writeVerifyKeyToFileError)
    } else{
      let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
      resolve(resolveData)
    }
  }
  
  @objc(readVerifyKeyFromFile:resolver:rejecter:)
  func readVerifyKeyFromFile(filePath: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.readVK( context_id , CString( Bundle.main.bundlePath + filePath).char() )
    if (rtn != 0){
      reject("event_failure","readVerifyKeyFromFile: \(rtn), \(lastmsg(contextId:context_id))", Errors.readVerifyKeyFromFileError)
    } else{
      let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
      resolve(resolveData)
    }
  }
  
  @objc(writeProofKeyToFile:resolver:rejecter:)
  func writeProofKeyToFile(filePath: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.writePK( context_id , CString( Bundle.main.bundlePath + filePath).char() )
    if (rtn != 0){
      reject("event_failure","writeProofKeyToFile: \(rtn), \(lastmsg(contextId:context_id))", Errors.writeProofKeyToFileError)
    } else{
      let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
      resolve(resolveData)
    }
  }
  
  @objc(readProofKeyFromFile:resolver:rejecter:)
  func readProofKeyFromFile(filePath: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.readPK( context_id , CString( Bundle.main.bundlePath + filePath).char() )
    if (rtn != 0){
      reject("event_failure","readProofKeyFromFile: \(rtn), \(lastmsg(contextId:context_id))", Errors.readProofKeyFromFileError)
    } else{
      let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
      resolve(resolveData)
    }
  }
  
  @objc(finalizeCircuit:rejecter:)
  func finalizeCircuit(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    CSnark.finalizeCircuit(context_id)
    resolve("Done")
  }
  
  @objc(setInput:hexValue:resolver:rejecter:)
  public func setInput(key : String , hexValue : String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock ) -> Void {
    let rtn = CSnark.updatePrimaryInputStr( context_id , CString(key).char() , CString(hexValue).char() )
    if rtn == 1 {
      reject("event_failure","setInput: \(rtn), \(lastmsg(contextId:context_id))", Errors.invalidPrimaryInputKey(invalidKey: key )
)
    } else {
      let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
      resolve(resolveData)
      
    }
  }
  
  
  @objc(setInputArray:arrayIdx:hexValue:resolver:rejecter:)
  public func setInputArray( key : String , arrayIdx : Int , hexValue : String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock ) -> Void {
    
    let rtn = CSnark.updatePrimaryInputArrayStr(
      context_id ,
      CString(key).char() ,
      Int32(arrayIdx) ,
      CString(hexValue).char())
    
    if rtn == 1 {
      reject("event_failure","setInputArray: \(rtn), \(lastmsg(contextId:context_id))", Errors.invalidPrimaryInputKey(invalidKey: key ))
    } else if rtn == 2 {
      reject("event_failure","setInputArray: \(rtn), \(lastmsg(contextId:context_id))", Errors.invalidPrimaryInputIndex(invalidIndex: arrayIdx , forKey : key ))
    } else {
      let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
      resolve(resolveData)
    }
    
  }

  @objc(setInputJsonString:resolver:rejecter:)
  public func setInputJsonString(  jsonStr: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    let rtn = CSnark.updatePrimaryInputFromJson(context_id, CString(jsonStr).char())
    
    if(rtn != 0) {
      reject("event_failure","setInputJsonString: \(rtn), \(lastmsg(contextId:context_id))", Errors.invalidPrimaryInputJson(invalidJson: jsonStr ))
    } else {
      let resolveData: [String: String] = ["lastmsg": lastmsg(contextId:context_id), "contextId": "\(context_id)", "rtn": "\(rtn)"]
      resolve(resolveData)
    }
  }
  
  func lastmsg( contextId: Int32) -> String {
    if let msgc : UnsafePointer<CChar> = getLastFunctionMsg( contextId ) {
      return String(cString: (msgc ))
    }else{
      return ""
    }
  }
  
  public enum Errors : Error {
    case unknownCircuitName
    case buildCircuitError
    case runSetupError
    case runProofError
    case writeVerifyKeyToFileError
    case readVerifyKeyFromFileError
    case writeProofKeyToFileError
    case readProofKeyFromFileError
    case serializeVerifyKeyError
    case serializeProofError
    case invalidPrimaryInputJson( invalidJson: String )
    case invalidPrimaryInputKey( invalidKey : String )
    case invalidPrimaryInputIndex( invalidIndex: Int , forKey : String )
  }
  
  class CString {
    
    var ns_string : NSString
    
    init(){
      ns_string = ""
    }
    
    init( _ base : String ) {
      ns_string = base as NSString
    }
    
    public func char() -> UnsafePointer<CChar>? {
      return UnsafePointer<CChar>( ns_string.utf8String )
    }
    
    public func str() -> String {
      return String(ns_string)
    }
    
  }
  
}

File: android/app/src/main/java/com/datatradeapp/LibsnarkModule.java
package com.datatradeapp;

import android.content.res.AssetManager;
import android.os.Environment;
import android.util.Log;

import androidx.annotation.NonNull;

import com.zKrypto.libSnarkJNI;

import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class LibsnarkModule extends ReactContextBaseJavaModule {
    private final ReactApplicationContext reactApplicationContext;
    private String crsDefaultPath = "";
    private int contextId = 0;
    private final Map<String, Integer> contextIdMap = new HashMap<String, Integer>(){{
        put("ZKlay", 0);
        put("ZKlay_nft", 0);
        put("GenTrade", 0);
    }};

    public libSnarkJNI snarkJNI = new libSnarkJNI();

    static int serializeFormatDefault = 1;
    static int serializeFormatCRV = 2;
    static int serializeFormatZKlay = 3;

    static int R1CS_GG = 1;
    static int R1CS_ROM_SE = 2;


    LibsnarkModule(ReactApplicationContext context) {
        super(context);
        this.reactApplicationContext = context;
        this.crsDefaultPath = context.getFilesDir().getAbsolutePath();
    }

    @Override
    public String getName() {
        return "LibsnarkModule";
    }

    @ReactMethod
    public void getContextId(String circuitName, Promise promise) {
        WritableMap resolveData = new WritableNativeMap();

        Integer retContextId = contextIdMap.get(circuitName);
        resolveData.putString("contextId", Integer.toString(retContextId == null ? 0 : retContextId));

        promise.resolve(resolveData);
    }

    @ReactMethod
    public void setContextId(String circuitName, Promise promise) {
        WritableMap resolveData = new WritableNativeMap();

        Integer retContextId = contextIdMap.get(circuitName);
        contextId = retContextId == null ? 0 : retContextId;
        resolveData.putString("contextId", Integer.toString(contextId));

        promise.resolve(resolveData);
    }


    @ReactMethod
    public void loggerTest(String name, Promise promise) {
        if (name.compareTo("") != 0) {
            promise.resolve(name);
        } else {
            Error error = new Error("Name: null");
            promise.reject("event_failure", "no str returned", error);
        }
    }

    @ReactMethod
    public void testMethod() {
        Log.d("info", "Print test message [libsnarkModule.java] ");
    }

    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        constants.put("DEFAULT_EVENT_NAME", "New Event");
        return constants;
    }

    @ReactMethod
    public void createCircuitContext(String circuitName, String treeHeight, String hashType, int serializeFormat, int ecSelection, Promise promise) {
        int _contextId = snarkJNI.createCircuitContext(circuitName, R1CS_GG, ecSelection, "", "", "");

        contextId = _contextId;
        contextIdMap.put(circuitName, _contextId);

        snarkJNI.serializeFormat(_contextId, serializeFormat);
        snarkJNI.assignCircuitArgument(_contextId, "treeHeight", treeHeight);
        snarkJNI.assignCircuitArgument(_contextId, "hashType", hashType);

        AssetManager assetManager = reactApplicationContext.getResources().getAssets();
        InputStream pkIs = null;
        InputStream vkIs = null;
        try {
            pkIs = assetManager.open("crs/" + circuitName + "_crs_pk.dat");
            vkIs = assetManager.open("crs/" + circuitName + "_crs_vk.dat");
            long pkFileSize = pkIs.available();
            long vkFileSize = vkIs.available();

            if (pkFileSize > 0 && vkFileSize > 0) {

                byte[] tempPkData = new byte[(int) pkFileSize];
                byte[] tempVkData = new byte[(int) vkFileSize];

                int pkLen = pkIs.read(tempPkData);
                int vkLen = vkIs.read(tempVkData);

                // copy CRS
                File crsDir = new File(crsDefaultPath + "/crs");

                if (!crsDir.exists()) {
                    crsDir.mkdir();
                }
                File pkOutFile = new File(crsDir + "/" + circuitName + "_crs_pk.dat");
                File vkOutFile = new File(crsDir + "/" + circuitName + "_crs_vk.dat");

                if (!pkOutFile.exists() || !vkOutFile.exists()) {
                    boolean a = pkOutFile.createNewFile();
                    boolean b = vkOutFile.createNewFile();
                }

                FileOutputStream pkFW = new FileOutputStream(pkOutFile, false);
                FileOutputStream vkFW = new FileOutputStream(vkOutFile, false);

                while (pkLen != -1) {
                    pkFW.write(tempPkData, 0, pkLen);
                    pkLen = pkIs.read(tempPkData);
                }
                pkFW.flush();
                pkFW.close();
                pkIs.close();

                while (vkLen != -1) {
                    vkFW.write(tempVkData, 0, vkLen);
                    vkLen = vkIs.read(tempVkData);
                }
                vkFW.flush();
                vkFW.close();
                vkIs.close();
            }
        } catch (Exception e) {
            Log.e("Exception", String.valueOf(e));
            promise.reject(e);
        }

        WritableMap resolveData = new WritableNativeMap();
        resolveData.putString("circuitName", circuitName);
        resolveData.putString("contextId", Integer.toString(_contextId));
        promise.resolve(resolveData);
    }

    @ReactMethod
    public void buildCircuit(Promise promise) {
        int rtn = snarkJNI.buildCircuit(contextId);
        if (rtn != 0) {
            promise.reject("event_failure", "buildCircuitError" + snarkJNI.getLastFunctionMsg(contextId));
        }

        WritableMap resolveData = new WritableNativeMap();
        resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
        resolveData.putString("contextId", Integer.toString(contextId));
        resolveData.putString("rtn", Integer.toString(rtn));
        promise.resolve(resolveData);
    }

    @ReactMethod
    public void runSetup(Promise promise) {
        int rtn = snarkJNI.runSetup(contextId);
        if (rtn != 0) {
            promise.reject("event_failure", "runSetupError" + snarkJNI.getLastFunctionMsg(contextId));
        }

        WritableMap resolveData = new WritableNativeMap();
        resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
        resolveData.putString("contextId", Integer.toString(contextId));
        resolveData.putString("rtn", Integer.toString(rtn));
        promise.resolve(resolveData);
    }

    @ReactMethod
    public void runProof(Promise promise) {
        int rtn = snarkJNI.runProof(contextId);
        if (rtn != 0) {
            promise.reject("event_failure", "runProofError" + snarkJNI.getLastFunctionMsg(contextId));
        }

        WritableMap resolveData = new WritableNativeMap();
        resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
        resolveData.putString("contextId", Integer.toString(contextId));
        resolveData.putString("rtn", Integer.toString(rtn));
        promise.resolve(resolveData);
    }

    @ReactMethod
    public void runVerify(Promise promise) {
        int rtn = snarkJNI.runVerify(contextId);
        if (rtn != 0) {
            promise.reject("event_failure", "runVerifyError" + snarkJNI.getLastFunctionMsg(contextId));
        }

        WritableMap resolveData = new WritableNativeMap();
        resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
        resolveData.putString("contextId", Integer.toString(contextId));
        resolveData.putString("rtn", Integer.toString(rtn));
        promise.resolve(resolveData);
    }

    @ReactMethod
    public void setProof(String proofJsonStr, Promise promise) {
        int rtn = snarkJNI.deSerializeProof(contextId, proofJsonStr);
        if (rtn != 0) {
            promise.reject("event_failure", "setProofError" + snarkJNI.getLastFunctionMsg(contextId));
        }

        WritableMap resolveData = new WritableNativeMap();
        resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
        resolveData.putString("contextId", Integer.toString(contextId));
        resolveData.putString("rtn", Integer.toString(rtn));
        promise.resolve(resolveData);
    }

    @ReactMethod
    public void getVerifyKeyJson(Promise promise) {
        String vk = snarkJNI.serializeVerifyKey(contextId);
        promise.resolve(vk);
    }

    @ReactMethod
    public void getProofJson(Promise promise) {
        String proofStr = snarkJNI.serializeProof(contextId);

        if (proofStr.isEmpty()) {
            promise.reject("event_failure", "serializeProofError" + snarkJNI.getLastFunctionMsg(contextId));
        } else {
            promise.resolve(proofStr);
        }
    }

    @ReactMethod
    public void writeVerifyKeyToFile(String filePath, Promise promise) {
        int rtn = snarkJNI.writeVK(contextId, crsDefaultPath + filePath);
        if (rtn != 0) {
            promise.reject("event_failure", "writeVerifyKeyToFileError in " + crsDefaultPath + filePath + " : " + snarkJNI.getLastFunctionMsg(contextId));
        } else {

            WritableMap resolveData = new WritableNativeMap();
            resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
            resolveData.putString("contextId", Integer.toString(contextId));
            resolveData.putString("rtn", Integer.toString(rtn));
            promise.resolve(resolveData);
        }
    }

    @ReactMethod
    public void readVerifyKeyFromFile(String filePath, Promise promise) {
        int rtn = snarkJNI.readVK(contextId, crsDefaultPath + filePath);
        if (rtn != 0) {

            promise.reject("event_failure", "readVerifyKeyFromFile in " + crsDefaultPath + filePath + " : " + snarkJNI.getLastFunctionMsg(contextId));
        } else {

            WritableMap resolveData = new WritableNativeMap();
            resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
            resolveData.putString("contextId", Integer.toString(contextId));
            resolveData.putString("rtn", Integer.toString(rtn));
            promise.resolve(resolveData);
        }
    }

    @ReactMethod
    public void writeProofKeyToFile(String filePath, Promise promise) {
        int rtn = snarkJNI.writePK(contextId, crsDefaultPath + filePath);
        if (rtn != 0) {
            promise.reject("event_failure", "writeProofKeyToFile in " + crsDefaultPath + filePath + " : " + snarkJNI.getLastFunctionMsg(contextId));
        } else {

            WritableMap resolveData = new WritableNativeMap();
            resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
            resolveData.putString("contextId", Integer.toString(contextId));
            resolveData.putString("rtn", Integer.toString(rtn));
            promise.resolve(resolveData);
        }
    }

    @ReactMethod
    public void readProofKeyFromFile(String filePath, Promise promise) {
        int rtn = snarkJNI.readPK(contextId, crsDefaultPath + filePath);
        if (rtn != 0) {
            promise.reject("event_failure", "readProofKeyFromFile in " + crsDefaultPath + filePath + " : " + snarkJNI.getLastFunctionMsg(contextId));
        } else {

            WritableMap resolveData = new WritableNativeMap();
            resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
            resolveData.putString("contextId", Integer.toString(contextId));
            resolveData.putString("rtn", Integer.toString(rtn));
            promise.resolve(resolveData);
        }
    }

    @ReactMethod
    public void finalizeCircuit(Promise promise) {
        snarkJNI.finalizeCircuit(contextId);
        promise.resolve("Done");
    }

    @ReactMethod
    public void setInput(String key, String hexValue, Promise promise) {
        // TODO
        promise.reject("event_failure", "setInputError: not supported in android");
    }


    @ReactMethod
    public void setInputArray(String key, int arrayIdx, String hexValue, Promise promise) {
        // TODO
        promise.reject("event_failure", "setInputArrayError: not supported in android");
    }

    @ReactMethod
    public void setInputJsonString(String jsonStr, Promise promise) {
        int rtn = snarkJNI.updatePrimaryInputFromJson(contextId, jsonStr);

        if (rtn != 0) {
            promise.reject("event_failure", "setInputJsonString" + snarkJNI.getLastFunctionMsg(contextId));
        } else {
            WritableMap resolveData = new WritableNativeMap();
            resolveData.putString("lastmsg", snarkJNI.getLastFunctionMsg(contextId));
            resolveData.putString("contextId", Integer.toString(contextId));
            resolveData.putString("rtn", Integer.toString(rtn));
            promise.resolve(resolveData);
        }
    }

    public String lastmsg(int contextId) {
        String msg = snarkJNI.getLastFunctionMsg(contextId);
        if (!msg.isEmpty()) {
            return msg;
        } else {
            return "";
        }
    }

}
